### ISSUE TITLE
mds downgrade from higher version to lower version

### ISSUE THREAD

**Issue by fengjiankui121** [rook/rook#8576]
**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**

**Expected behavior:**
not allowed mds downgrade

**How to reproduce it (minimal and precise):**

step 1: install ceph with higher version
step 2: upgrade ceph with lower version
step 3: Wait for the osd deployment to complete, restart the operator
step 4: mds will downgrade from higher version to lower version

**File(s) to submit**:

* Cluster CR (custom resource), typically called `cluster.yaml`, if necessary
* Operator's logs, if necessary
* Crashing pod(s) logs, if necessary

 To get logs, use `kubectl -n <namespace> logs <pod name>`
When pasting logs, always surround them with backticks or use the `insert code` button from the Github UI.
Read [Github documentation if you need help](https://help.github.com/en/articles/creating-and-highlighting-code-blocks).

**Environment**:
* OS (e.g. from /etc/os-release):
* Kernel (e.g. `uname -a`):
* Cloud provider or hardware configuration:
* Rook version (use `rook version` inside of a Rook Pod):
* Storage backend version (e.g. for ceph do `ceph -v`):
* Kubernetes version (use `kubectl version`):
* Kubernetes cluster type (e.g. Tectonic, GKE, OpenShift):
* Storage backend status (e.g. for Ceph use `ceph health` in the [Rook Ceph toolbox](https://rook.io/docs/rook/master/ceph-toolbox.html)):

**Comment by travisn** [comment]
@fengjiankui121 If you don't restart the operator do you see the same behavior? The operator intends to prevent the ceph downgrades, but there may be an issue when the operator is restarted.

**Comment by fengjiankui121** [comment]
@travisn The same phenomenon will occur if other operations cause the operator to re-run. For example, cephcluster configuration changes

**Comment by fengjiankui121** [comment]
@travisn This problem is mainly due to the inconsistency of the version of the ceph component, which triggers the upgrade of mds. The relevant code is as follows:

rook\pkg\operator\ceph\cluster\version.go

if numberOfCephVersions > 1 {
  // let's return immediately
  logger.Warningf("it looks like we have more than one ceph version running. triggering upgrade. %+v:", runningVersions.Overall)
  return true, nil
}

**Comment by fengjiankui121** [comment]
@travisn OSD, monitor, etc. are allowed to upgrade to lower versions, but mds is not allowed, which will lead to inconsistent versions of ceph components, and inconsistent versions of ceph components will trigger the upgrade of mds

**Comment by travisn** [comment]
@fengjiankui121 To be clear, are you seeing that the mds is *not* updated during a downgraded Ceph version? But if you restart the operator, the mds *is* updated to the downgraded version? And the bug is that you expect the mds to downgrade without the operator restart?

**Comment by fengjiankui121** [comment]
@travisn  yes, it is

**Comment by travisn** [comment]
Ok, sounds like we just need to relax the check and allow the mds downgrade. It's not really supported to downgrade, but the reality is that sometimes it is better to risk the downgrade than to stay in a broken state after an upgrade

**Comment by sp98** [comment]
@fengjiankui121  Need help the reproduce this behavior.  Steps I followed are below: 

- Deploy rook-ceph (master branch) with ceph 16.2.5. 
- Create `cluster/examples/kubernetes/ceph/filesystem-test.yaml` 
- Downgrade ceph version to `image: quay.io/ceph/ceph:v16.2.4` 
- Wait for downgrade to complete. 

Before downgrade: 
```
 versions:
      mds:
        ceph version 16.2.5 (0883bdea7337b95e4b611c768c0279868462204a) pacific (stable): 2
      mgr:
        ceph version 16.2.5 (0883bdea7337b95e4b611c768c0279868462204a) pacific (stable): 1
      mon:
        ceph version 16.2.5 (0883bdea7337b95e4b611c768c0279868462204a) pacific (stable): 3
      osd:
        ceph version 16.2.5 (0883bdea7337b95e4b611c768c0279868462204a) pacific (stable): 1
      overall:
        ceph version 16.2.5 (0883bdea7337b95e4b611c768c0279868462204a) pacific (stable): 7
```

After downgrade :

```
    versions:
      mds:
        ceph version 16.2.4 (3cbe25cde3cfa028984618ad32de9edc4c1eaed0) pacific (stable): 2
      mgr:
        ceph version 16.2.4 (3cbe25cde3cfa028984618ad32de9edc4c1eaed0) pacific (stable): 1
      mon:
        ceph version 16.2.4 (3cbe25cde3cfa028984618ad32de9edc4c1eaed0) pacific (stable): 3
      osd:
        ceph version 16.2.4 (3cbe25cde3cfa028984618ad32de9edc4c1eaed0) pacific (stable): 1
      overall:
        ceph version 16.2.4 (3cbe25cde3cfa028984618ad32de9edc4c1eaed0) pacific (stable): 7
``

**Comment by fengjiankui121** [comment]
From the rook code, it can be found that mds is not allowed to be upgraded, unless the versions between the ceph components are inconsistent, the relevant code is as follows：

if numberOfCephVersions > 1 {
// let's return immediately
  logger.Warningf("it looks like we have more than one ceph version running. triggering upgrade. %+v:", runningVersions.Overall)
return true, nil
}

......
......

if cephver.IsInferior(imageSpecVersion, clusterRunningVersion) {
    return true, errors.Errorf("image spec version %s is lower than the running cluster version %s, downgrading is not supported", imageSpecVersion.String(), clusterRunningVersion.String())}

**Comment by fengjiankui121** [comment]
@sp98  From the rook code, it can be found that mds is not allowed to be upgraded, unless the versions between the ceph components are inconsistent, the relevant code is as follows：

if numberOfCephVersions > 1 {
// let's return immediately
  logger.Warningf("it looks like we have more than one ceph version running. triggering upgrade. %+v:", runningVersions.Overall)
return true, nil
}

......
......

if cephver.IsInferior(imageSpecVersion, clusterRunningVersion) {
    return true, errors.Errorf("image spec version %s is lower than the running cluster version %s, downgrading is not supported", imageSpecVersion.String(), clusterRunningVersion.String())}

**Comment by sp98** [comment]
Able to reproduce this issue. 

```
  versions:
      mds:
        ceph version 16.2.5 (0883bdea7337b95e4b611c768c0279868462204a) pacific (stable): 2
      mgr:
        ceph version 16.2.4 (3cbe25cde3cfa028984618ad32de9edc4c1eaed0) pacific (stable): 1
      mon:
        ceph version 16.2.4 (3cbe25cde3cfa028984618ad32de9edc4c1eaed0) pacific (stable): 3
      osd:
        ceph version 16.2.4 (3cbe25cde3cfa028984618ad32de9edc4c1eaed0) pacific (stable): 1
      overall:
        ceph version 16.2.4 (3cbe25cde3cfa028984618ad32de9edc4c1eaed0) pacific (stable): 5
        ceph version 16.2.5 (0883bdea7337b95e4b611c768c0279868462204a) pacific (stable): 2
```

Discussed about this issue in the huddle. General consensus is to allow users to downgrade. See [comment](https://github.com/rook/rook/issues/8576#issuecomment-919460199) 

Rook ( and also ceph) does not support downgrades. 

Waiting for a major refactoring [PR](https://github.com/rook/rook/pull/8607).  I'll test this issue again after that PR is merged.

**Comment by travisn** [comment]
This was actually fixed by #9098 in v1.7.7
